File: toyhana-front/src/utils/phone.ts
/**
 * Нормализация и форматирование казахстанских номеров.
 * Бэк принимает только +77XXXXXXXXX.
 */

/** Привести любой ввод к +77XXXXXXXXX или вернуть null, если не получилось. */
export function normalizeKzPhone(raw: string): string | null {
  const digits = (raw || '').replace(/\D/g, '');
  let d = digits;
  // 8XXXXXXXXXX -> 7XXXXXXXXXX
  if (d.length === 11 && d.startsWith('8')) d = '7' + d.slice(1);
  if (d.length !== 11 || !d.startsWith('77')) return null;
  return '+' + d;
}

/** Формат для показа: +7 (701) 234 56 78 */
export function formatKzPhoneDisplay(phone: string): string {
  const d = phone.replace(/\D/g, '');
  if (d.length !== 11 || !d.startsWith('77')) return phone;
  return `+7 (${d.slice(1, 4)}) ${d.slice(4, 7)} ${d.slice(7, 9)} ${d.slice(9, 11)}`;
}

/**
 * Привести ввод пользователя к красивому виду по мере набора.
 * Пример: '87011' -> '+7 (701) 1'
 * Всегда возвращает строку, которая начинается с '+7 '.
 */
export function maskPhoneInput(raw: string): string {
  let digits = (raw || '').replace(/\D/g, '');
  if (digits.startsWith('8')) digits = '7' + digits.slice(1);
  if (!digits.startsWith('7')) digits = '7' + digits;
  digits = digits.slice(0, 11);

  const rest = digits.slice(1);              // без ведущей 7
  const p1 = rest.slice(0, 3);
  const p2 = rest.slice(3, 6);
  const p3 = rest.slice(6, 8);
  const p4 = rest.slice(8, 10);

  let out = '+7';
  if (p1) out += ` (${p1}`;
  if (p1.length === 3) out += ')';
  if (p2) out += ` ${p2}`;
  if (p3) out += ` ${p3}`;
  if (p4) out += ` ${p4}`;
  return out;
}

File: toyhana-back/libs/phone.py
"""
Работа с номером телефона.
Формат хранения и передачи по API: +7XXXXXXXXXX (12 символов, только цифры после +7).
"""
import re

from handlers.h01_errors import WarnException


def normalize_phone(raw: str) -> str:
    """
    Приводит номер к формату +7XXXXXXXXXX.
    Принимает варианты:
        87011234567
        +77011234567
        +7 (701) 123-45-67
        8 701 123 45 67
    Если номер некорректный — бросает WarnException(400).
    """
    if raw is None:
        raise WarnException(400, 'Номер телефона обязателен')

    # Оставляем только цифры
    digits = re.sub(r'\D', '', str(raw))

    # 8XXXXXXXXXX -> 7XXXXXXXXXX (11 цифр, начинается с 8)
    if len(digits) == 11 and digits.startswith('8'):
        digits = '7' + digits[1:]

    # Должно получиться 11 цифр, начинающихся с 77 (казахстанский мобильный префикс).
    # Российские +7 (начинающиеся на 79, 74, ...) не принимаем — у нас сервис для КЗ.
    if len(digits) != 11 or not digits.startswith('77'):
        raise WarnException(400, 'Некорректный номер телефона. Формат: +77XXXXXXXXX')

    return '+' + digits
